### ISSUE TITLE
incompatible_no_transitive_loads: Forbid macros loaded transitively

### ISSUE THREAD

**Issue by laurentlb** [bazelbuild/bazel#5636]
When a bzl file contains a load, e.g.
`load("//some:file.bzl", "fct")`
it implicitly exports `fct`.

This is problematic for many reasons:
- Exports should be explicit
- Other languages don't do that
- Removing a load can break other files
- So it means we can't have a tool to safely remove unused loads

Unfortunately, this is a breaking change.

There are two ways to fix it.

1. Update the load site and use the label of the file where the symbol is really declared
1. Update the loaded file. To explicitly re-export the symbol `foo`, use:

```python
load("...", _foo = "foo")

foo = _foo
```

**Comment by ittaiz** [comment]
+1 on this change. Current behavior is surprising.

**Comment by laurentlb** [comment]
After running on the CI (https://buildkite.com/bazel/bazel-at-release-plus-incompatible-flags/builds/56), here's the list of projects that fail with the flag:
 - Android Testing
 - bazel-toolchains
 - Gerrit
 - rules_closure
 - rules_foreign_cc
 - rules_scala
 - Tensorflow

**Comment by laurentlb** [comment]
I think so. I'll update the list after a new CI run.

**Comment by philwo** [comment]
This flag was not flipped in time for the Bazel 0.23.0 release and will thus be postponed to Bazel 0.24.0.

**Comment by laurentlb** [comment]
Still blocked by: Android Testing, Bazel integration testing, Bazel Remote Cache, Bazel Watcher, Gerrit, Tensorflow, rules_webtesting.

**Comment by katre** [comment]
This issue was tagged as "breaking-change-0.24" but does not appear ready to be flipped in the [0.24.0 release](https://github.com/bazelbuild/bazel/issues/6968). If this is incorrect please comment on that issue and discuss with me.

**Comment by laurentlb** [comment]
It looks like we can now flip the flag: https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/37

**Comment by dhalperi** [comment]
@laurentlb - is there any good debugging advice for how to fix/debug related issues in generated code?

I'm trying to make https://github.com/pubref/rules_maven work with this flag. Fix 2 in the original comment worked for the bzl files that are there. But I can't figure out how to debug issues in generated files -- I just get warnings about cycles in the graph -- and it seems bazel doesn't materialize the generated files somewhere I can look at them (and mutate them).

E.g., in https://github.com/batfish/batfish @ master using bazel from yesterday HEAD,

```
$ bazel build //... --incompatible_no_transitive_loads=true        
INFO: Invocation ID: 0a92899e-0907-4642-877e-ae1644a13cf9
ERROR: Failed to load Starlark extension '@antlr4_runtime//:rules.bzl'.
It usually happens when the repository is not defined prior to being used.
This could either mean you have to add the 'antlr4_runtime' repository with a statement like `http_archive` in your WORKSPACE file (note that transitive dependencies are not added automatically), or the repository 'antlr4_runtime' was defined too late in your WORKSPACE file.
ERROR: cycles detected during target parsing
INFO: Elapsed time: 0.084s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
    Fetching @antlr4_runtime; Restarting.
```

but no files are materialized (and the bazel-foo symlinks aren't even created) for me to look at.

**Comment by jin** [comment]
Tensorflow at HEAD requires disabling this flag in Bazel HEAD in order to build.

**Comment by laurentlb** [comment]
This was temporary rolled back. I plan to roll it forward next week (to be included in Bazel 0.25).

**Comment by ittaiz** [comment]
@laurentlb to verify- the fact that this issue is closed means new code can use `load` "regularly" and not have to worry about it being exported, right?

**Comment by laurentlb** [comment]
Yes.
We close an issue about an incompatible flag when the change is enabled by default at head. So you can expect all Bazel releases after March 2019 to behave correctly in this regard.
